atlasdb-feign/src/main/java/com/palantir/atlasdb/http/FailoverFeignTarget.java

```diff
@@ -57,25 +62,30 @@
 
     private final AtomicLong failuresSinceLastSwitch = new AtomicLong();
     private final AtomicLong numSwitches = new AtomicLong();
-    private final AtomicLong startTimeOfFastFailover = new AtomicLong();
+    private final AtomicReference<Optional<FastFailoverState>> fastFailoverState
```

Comment:
In your PR you say that each request gets a copy of FailoverFeignTarget, I don't see that? I think we absolutely need the atomics here.

The concurrency setup in this class is a bit special but it doesn't seem immediately off.

Reply:
I said this because FailoverFeignTarget implements Retryer, where the javadocs say

```
 * Cloned for each invocation to {@link Client#execute(Request, feign.Request.Options)}.
 * Implementations may keep state to determine if retry operations should continue or not.
```

Though I don't understand feign internals sufficiently well to be 100% sure that removing this is safe.

Reply:
We spoke offline - it's not actually safe to remove this. Clone() is called, but our impl of clone() does a return this;